File: frontend/src/lib/utils/access-policy.ts
import type { AccessSurface, PermissionsManifest, User } from '$lib/types/auth';
import { GLOBAL_SCOPE, SUDO_PERMISSION } from '$lib/types/auth';

type SurfaceIndex = Map<string, AccessSurface>;

let cachedSurfacesInternal: AccessSurface[] | undefined;
let cachedSurfaceIndexInternal: SurfaceIndex | undefined;

export function canReachAccessSurface(
	manifest: PermissionsManifest | null | undefined,
	surfaceId: string | undefined,
	user: User | null | undefined,
	selectedEnvId?: string
): boolean {
	if (!surfaceId || !user) return false;
	const surfaces = manifest?.accessSurfaces;
	if (!surfaces?.length) return false;

	return canReachSurfaceInternal(getSurfaceIndexInternal(surfaces), surfaceId, user, selectedEnvId, new Set());
}

export function canReachAccessSurfaceUrl(
	manifest: PermissionsManifest | null | undefined,
	url: string,
	user: User | null | undefined,
	selectedEnvId?: string
): boolean {
	const surface = manifest?.accessSurfaces?.find((entry) => entry.url === url);
	if (!surface) return false;
	return canReachAccessSurface(manifest, surface.id, user, selectedEnvId);
}

export function getRouteAccessSurfaces(manifest: PermissionsManifest | null | undefined): AccessSurface[] {
	return (manifest?.accessSurfaces ?? [])
		.filter((surface) => !!surface.url)
		.slice()
		.sort((a, b) => (b.url?.length ?? 0) - (a.url?.length ?? 0));
}

export function getFallbackAccessSurfaces(manifest: PermissionsManifest | null | undefined): AccessSurface[] {
	return (manifest?.accessSurfaces ?? [])
		.filter((surface) => !!surface.url && !!surface.fallbackOrder)
		.slice()
		.sort((a, b) => (a.fallbackOrder ?? 0) - (b.fallbackOrder ?? 0));
}

export function pathMatchesAccessSurface(path: string, surface: AccessSurface): boolean {
	const template = surface.url;
	if (!template) return false;

	const pathSegments = splitPathInternal(path);
	const templateSegments = splitPathInternal(template);
	if (templateSegments.length > pathSegments.length) return false;

	for (let i = 0; i < templateSegments.length; i++) {
		const expected = templateSegments[i];
		const actual = pathSegments[i];
		if (!expected || !actual) return false;
		if (isTemplateSegmentInternal(expected)) continue;
		if (expected !== actual) return false;
	}

	return true;
}

function canReachSurfaceInternal(
	surfaces: SurfaceIndex,
	surfaceId: string,
	user: User,
	selectedEnvId: string | undefined,
	visiting: Set<string>
): boolean {
	if (visiting.has(surfaceId)) return false;

	const surface = surfaces.get(surfaceId);
	if (!surface) return false;

	if (surface.accessMode === 'any-child') {
		visiting.add(surfaceId);
		const reachable = (surface.children ?? []).some((childId) =>
			canReachSurfaceInternal(surfaces, childId, user, selectedEnvId, visiting)
		);
		visiting.delete(surfaceId);
		return reachable;
	}

	const permissions = surface.permissions ?? [];
	if (permissions.length === 0) return false;

	if (surface.matchMode === 'all-of') {
		return permissions.every((permission) => hasPermissionForScopeInternal(user, permission, surface.scopeMode, selectedEnvId));
	}

	return permissions.some((permission) => hasPermissionForScopeInternal(user, permission, surface.scopeMode, selectedEnvId));
}

function hasPermissionForScopeInternal(
	user: User,
	permission: string,
	scopeMode: AccessSurface['scopeMode'],
	selectedEnvId: string | undefined
): boolean {
	const global = user.permissionsByEnv?.[GLOBAL_SCOPE] ?? [];
	if (global.includes(SUDO_PERMISSION)) return true;

	switch (scopeMode) {
		case 'global-only':
			return global.includes(permission);
		case 'selected-env-plus-global':
			return permissionsForEnvInternal(user, selectedEnvId).has(permission);
		case 'any-effective-scope':
			return Object.values(user.permissionsByEnv ?? {}).some((permissions) => permissions.includes(permission));
		default:
			return false;
	}
}

function permissionsForEnvInternal(user: User, envId: string | undefined): Set<string> {
	const out = new Set<string>();
	const global = user.permissionsByEnv?.[GLOBAL_SCOPE];
	if (global) for (const permission of global) out.add(permission);
	if (envId && envId !== GLOBAL_SCOPE) {
		const scoped = user.permissionsByEnv?.[envId];
		if (scoped) for (const permission of scoped) out.add(permission);
	}
	return out;
}

function buildSurfaceIndexInternal(surfaces: AccessSurface[]): SurfaceIndex {
	return new Map(surfaces.map((surface) => [surface.id, surface]));
}

function getSurfaceIndexInternal(surfaces: AccessSurface[]): SurfaceIndex {
	if (cachedSurfacesInternal === surfaces && cachedSurfaceIndexInternal) {
		return cachedSurfaceIndexInternal;
	}

	cachedSurfacesInternal = surfaces;
	cachedSurfaceIndexInternal = buildSurfaceIndexInternal(surfaces);
	return cachedSurfaceIndexInternal;
}

function splitPathInternal(path: string): string[] {
	const withoutQuery = path.split('?')[0] ?? '';
	const withoutHash = withoutQuery.split('#')[0] ?? '';
	return withoutHash.split('/').filter(Boolean);
}

function isTemplateSegmentInternal(segment: string): boolean {
	return segment.startsWith(':') || (segment.startsWith('{') && segment.endsWith('}'));
}

File: backend/pkg/authz/access_policy.go
package authz

const (
	AccessSurfaceKindRoute             = "route"
	AccessSurfaceKindSettingsCategory  = "settings-category"
	AccessSurfaceKindCustomizeCategory = "customize-category"
	AccessSurfaceKindLanding           = "landing"

	AccessModePermissions = "permissions"
	AccessModeAnyChild    = "any-child"

	AccessMatchModeAnyOf = "any-of"
	AccessMatchModeAllOf = "all-of"

	AccessScopeModeGlobalOnly            = "global-only"
	AccessScopeModeSelectedEnvPlusGlobal = "selected-env-plus-global"
	AccessScopeModeAnyEffectiveScope     = "any-effective-scope"
)

// AccessSurface describes one frontend-visible surface whose reachability is
// derived from backend-owned permission metadata. This is decision metadata for
// UX gating; backend middleware and service checks remain authoritative.
type AccessSurface struct {
	ID            string
	Kind          string
	URL           string
	Label         string
	AccessMode    string
	MatchMode     string
	ScopeMode     string
	Permissions   []string
	Children      []string
	FallbackOrder int
}

var accessSurfacesInternal = []AccessSurface{
	landingSurfaceInternal("landing.customize", "/customize", "Customize", []string{
		"customize.category.templates",
		"customize.category.registries",
		"customize.category.variables",
		"customize.category.git-repositories",
	}, 40),
	landingSurfaceInternal("landing.settings", "/settings", "Settings", []string{
		"settings.category.activity",
		"settings.category.apikeys",
		"settings.category.appearance",
		"settings.category.authentication",
		"settings.category.build",
		"settings.category.jobschedule",
		"settings.category.notifications",
		"settings.category.roles",
		"settings.category.timeouts",
		"settings.category.users",
		"settings.category.webhooks",
		"settings.category.diagnostics",
	}, 120),

	routeSurfaceInternal("route.dashboard", "/dashboard", "Dashboard", AccessScopeModeSelectedEnvPlusGlobal, []string{PermDashboardRead}, 10),
	routeSurfaceInternal("route.projects", "/projects", "Projects", AccessScopeModeSelectedEnvPlusGlobal, []string{PermProjectsList, PermProjectsRead}, 30),
	routeSurfaceInternal("route.environments", "/environments", "Environments", AccessScopeModeGlobalOnly, []string{PermEnvironmentsList, PermEnvironmentsRead}, 130),
	routeSurfaceInternal("route.environments.gitops", "/environments/{id}/gitops", "GitOps Syncs", AccessScopeModeSelectedEnvPlusGlobal, []string{PermGitOpsList, PermGitOpsRead}, 0),
	routeSurfaceInternal("route.containers", "/containers", "Containers", AccessScopeModeSelectedEnvPlusGlobal, []string{PermContainersList, PermContainersRead}, 20),
	routeSurfaceInternal("route.images", "/images", "Images", AccessScopeModeSelectedEnvPlusGlobal, []string{PermImagesList, PermImagesRead}, 50),
	routeSurfaceInternal("route.images.builds", "/images/builds", "Builds", AccessScopeModeSelectedEnvPlusGlobal, []string{PermImagesBuild}, 0),
	routeSurfaceInternal("route.images.vulnerabilities", "/images/vulnerabilities", "Vulnerabilities", AccessScopeModeSelectedEnvPlusGlobal, []string{PermVulnsRead}, 0),
	routeSurfaceInternal("route.updates", "/updates", "Image Updates", AccessScopeModeSelectedEnvPlusGlobal, []string{PermImageUpdatesRead}, 0),
	routeSurfaceInternal("route.networks", "/networks", "Networks", AccessScopeModeSelectedEnvPlusGlobal, []string{PermNetworksList, PermNetworksRead}, 70),
	routeSurfaceInternal("route.ports", "/ports", "Ports", AccessScopeModeSelectedEnvPlusGlobal, []string{PermContainersList}, 0),
	routeSurfaceInternal("route.networks.topology", "/networks/topology", "Network Topology", AccessScopeModeSelectedEnvPlusGlobal, []string{PermNetworksRead}, 0),
	routeSurfaceInternal("route.volumes", "/volumes", "Volumes", AccessScopeModeSelectedEnvPlusGlobal, []string{PermVolumesList, PermVolumesRead}, 60),
	routeSurfaceInternal("route.swarm", "/swarm", "Swarm", AccessScopeModeSelectedEnvPlusGlobal, []string{PermSwarmRead}, 0),
	routeSurfaceInternal("route.swarm.services", "/swarm/services", "Services", AccessScopeModeSelectedEnvPlusGlobal, []string{PermSwarmServices}, 80),
	routeSurfaceInternal("route.swarm.nodes", "/swarm/nodes", "Nodes", AccessScopeModeSelectedEnvPlusGlobal, []string{PermSwarmNodes}, 0),
	routeSurfaceInternal("route.swarm.tasks", "/swarm/tasks", "Tasks", AccessScopeModeSelectedEnvPlusGlobal, []string{PermSwarmRead}, 0),
	routeSurfaceInternal("route.swarm.stacks", "/swarm/stacks", "Stacks", AccessScopeModeSelectedEnvPlusGlobal, []string{PermSwarmStacks}, 90),
	routeSurfaceInternal("route.swarm.cluster", "/swarm/cluster", "Cluster", AccessScopeModeSelectedEnvPlusGlobal, []string{PermSwarmRead}, 100),
	routeSurfaceInternal("route.swarm.configs", "/swarm/configs", "Configs", AccessScopeModeSelectedEnvPlusGlobal, []string{PermSwarmConfigs}, 0),
	routeSurfaceInternal("route.swarm.secrets", "/swarm/secrets", "Secrets", AccessScopeModeSelectedEnvPlusGlobal, []string{PermSwarmSecrets}, 0),
	routeSurfaceInternal("route.events", "/events", "Events", AccessScopeModeGlobalOnly, []string{PermEventsRead}, 110),

	settingsCategorySurfaceInternal("activity", "/settings/activity", "Activity", AccessScopeModeGlobalOnly, []string{PermSettingsRead}),
	settingsCategorySurfaceInternal("apikeys", "/settings/api-keys", "API Keys", AccessScopeModeGlobalOnly, []string{PermApiKeysList, PermApiKeysRead}),
	settingsCategorySurfaceInternal("appearance", "/settings/appearance", "Appearance", AccessScopeModeGlobalOnly, []string{PermSettingsRead}),
	settingsCategorySurfaceInternal("authentication", "/settings/authentication", "Authentication", AccessScopeModeGlobalOnly, []string{PermSettingsRead}),
	settingsCategorySurfaceInternal("build", "/settings/builds", "Builds", AccessScopeModeGlobalOnly, []string{PermSettingsRead}),
	settingsCategorySurfaceInternal("jobschedule", "/settings/jobs", "Job Schedule", AccessScopeModeSelectedEnvPlusGlobal, []string{PermJobsManage}),
	settingsCategorySurfaceInternal("notifications", "/settings/notifications", "Notifications", AccessScopeModeSelectedEnvPlusGlobal, []string{PermNotificationsManage}),
	settingsCategorySurfaceInternal("roles", "/settings/roles", "Roles", AccessScopeModeGlobalOnly, []string{PermRolesList, PermRolesRead}),
	settingsCategorySurfaceInternal("timeouts", "/settings/timeouts", "Timeouts", AccessScopeModeGlobalOnly, []string{PermSettingsRead}),
	settingsCategorySurfaceInternal("users", "/settings/users", "Users", AccessScopeModeGlobalOnly, []string{PermUsersList, PermUsersRead}),
	settingsCategorySurfaceInternal("webhooks", "/settings/webhooks", "Webhooks", AccessScopeModeSelectedEnvPlusGlobal, []string{PermWebhooksList}),
	settingsCategorySurfaceInternal("diagnostics", "/settings/diagnostics", "Diagnostics", AccessScopeModeGlobalOnly, []string{PermDiagnosticsRead}),

	customizeCategorySurfaceInternal("templates", "/customize/templates", "Templates", []string{PermCustomizeManage, PermTemplatesList, PermTemplatesRead}),
	customizeCategorySurfaceInternal("registries", "/customize/registries", "Container Registries", []string{PermCustomizeManage, PermRegistriesList, PermRegistriesRead}),
	customizeCategorySurfaceInternal("variables", "/customize/variables", "Variables", []string{PermCustomizeManage, PermTemplatesRead}),
	customizeCategorySurfaceInternal("git-repositories", "/customize/git-repositories", "Git Repositories", []string{PermCustomizeManage, PermGitReposList, PermGitReposRead}),
}

var accessSurfacesByIDInternal = buildAccessSurfaceIndexInternal(accessSurfacesInternal)

// AccessSurfaces returns a defensive copy of every backend-owned access
// surface in stable evaluation order.
func AccessSurfaces() []AccessSurface {
	out := make([]AccessSurface, len(accessSurfacesInternal))
	for i := range accessSurfacesInternal {
		out[i] = copyAccessSurfaceInternal(accessSurfacesInternal[i])
	}
	return out
}

// AccessSurfaceByID returns a defensive copy of the requested access surface.
func AccessSurfaceByID(id string) (AccessSurface, bool) {
	surface, ok := accessSurfacesByIDInternal[id]
	if !ok {
		return AccessSurface{}, false
	}
	return copyAccessSurfaceInternal(surface), true
}

// AccessSurfacesByKind returns all surfaces of the requested kind.
func AccessSurfacesByKind(kind string) []AccessSurface {
	out := make([]AccessSurface, 0)
	for _, surface := range accessSurfacesInternal {
		if surface.Kind == kind {
			out = append(out, copyAccessSurfaceInternal(surface))
		}
	}
	return out
}

// CanAccessSurface evaluates backend-owned UI reachability metadata for the
// caller. It is for advisory UX only; middleware and handlers still enforce
// permissions on actual API requests.
func CanAccessSurface(ps *PermissionSet, surfaceID, selectedEnvID string) bool {
	return canAccessSurfaceInternal(ps, surfaceID, selectedEnvID, make(map[string]struct{}))
}

// CanAccessSettingsCategory reports whether a settings category is reachable
// for the selected environment.
func CanAccessSettingsCategory(ps *PermissionSet, categoryID, selectedEnvID string) bool {
	return CanAccessSurface(ps, "settings.category."+categoryID, selectedEnvID)
}

// CanAccessCustomizeCategory reports whether a customize category is reachable.
func CanAccessCustomizeCategory(ps *PermissionSet, categoryID, selectedEnvID string) bool {
	return CanAccessSurface(ps, "customize.category."+categoryID, selectedEnvID)
}

func canAccessSurfaceInternal(ps *PermissionSet, surfaceID, selectedEnvID string, visiting map[string]struct{}) bool {
	if ps == nil {
		return false
	}
	if _, ok := visiting[surfaceID]; ok {
		return false
	}
	surface, ok := accessSurfacesByIDInternal[surfaceID]
	if !ok {
		return false
	}

	switch surface.AccessMode {
	case AccessModeAnyChild:
		visiting[surfaceID] = struct{}{}
		defer delete(visiting, surfaceID)
		for _, childID := range surface.Children {
			if canAccessSurfaceInternal(ps, childID, selectedEnvID, visiting) {
				return true
			}
		}
		return false
	case AccessModePermissions:
		return allowsSurfacePermissionsInternal(ps, surface, selectedEnvID)
	default:
		return false
	}
}

func allowsSurfacePermissionsInternal(ps *PermissionSet, surface AccessSurface, selectedEnvID string) bool {
	if len(surface.Permissions) == 0 {
		return false
	}

	switch surface.MatchMode {
	case AccessMatchModeAllOf:
		for _, perm := range surface.Permissions {
			if !allowsPermissionForScopeModeInternal(ps, perm, surface.ScopeMode, selectedEnvID) {
				return false
			}
		}
		return true
	case AccessMatchModeAnyOf:
		for _, perm := range surface.Permissions {
			if allowsPermissionForScopeModeInternal(ps, perm, surface.ScopeMode, selectedEnvID) {
				return true
			}
		}
		return false
	default:
		return false
	}
}

func allowsPermissionForScopeModeInternal(ps *PermissionSet, perm, scopeMode, selectedEnvID string) bool {
	switch scopeMode {
	case AccessScopeModeGlobalOnly:
		return ps.Allows(perm, "")
	case AccessScopeModeSelectedEnvPlusGlobal:
		return ps.Allows(perm, selectedEnvID)
	case AccessScopeModeAnyEffectiveScope:
		if ps.Allows(perm, "") {
			return true
		}
		for envID := range ps.PerEnv {
			if ps.Allows(perm, envID) {
				return true
			}
		}
		return false
	default:
		return false
	}
}

func buildAccessSurfaceIndexInternal(surfaces []AccessSurface) map[string]AccessSurface {
	out := make(map[string]AccessSurface, len(surfaces))
	for _, surface := range surfaces {
		out[surface.ID] = surface
	}
	return out
}

func copyAccessSurfaceInternal(surface AccessSurface) AccessSurface {
	surface.Permissions = append([]string(nil), surface.Permissions...)
	surface.Children = append([]string(nil), surface.Children...)
	return surface
}

func routeSurfaceInternal(id, url, label, scopeMode string, permissions []string, fallbackOrder int) AccessSurface {
	return AccessSurface{
		ID:            id,
		Kind:          AccessSurfaceKindRoute,
		URL:           url,
		Label:         label,
		AccessMode:    AccessModePermissions,
		MatchMode:     AccessMatchModeAnyOf,
		ScopeMode:     scopeMode,
		Permissions:   append([]string(nil), permissions...),
		FallbackOrder: fallbackOrder,
	}
}

func settingsCategorySurfaceInternal(categoryID, url, label, scopeMode string, permissions []string) AccessSurface {
	return AccessSurface{
		ID:          "settings.category." + categoryID,
		Kind:        AccessSurfaceKindSettingsCategory,
		URL:         url,
		Label:       label,
		AccessMode:  AccessModePermissions,
		MatchMode:   AccessMatchModeAnyOf,
		ScopeMode:   scopeMode,
		Permissions: append([]string(nil), permissions...),
	}
}

func customizeCategorySurfaceInternal(categoryID, url, label string, permissions []string) AccessSurface {
	return AccessSurface{
		ID:          "customize.category." + categoryID,
		Kind:        AccessSurfaceKindCustomizeCategory,
		URL:         url,
		Label:       label,
		AccessMode:  AccessModePermissions,
		MatchMode:   AccessMatchModeAnyOf,
		ScopeMode:   AccessScopeModeGlobalOnly,
		Permissions: append([]string(nil), permissions...),
	}
}

func landingSurfaceInternal(id, url, label string, children []string, fallbackOrder int) AccessSurface {
	return AccessSurface{
		ID:            id,
		Kind:          AccessSurfaceKindLanding,
		URL:           url,
		Label:         label,
		AccessMode:    AccessModeAnyChild,
		MatchMode:     AccessMatchModeAnyOf,
		ScopeMode:     AccessScopeModeSelectedEnvPlusGlobal,
		Children:      append([]string(nil), children...),
		FallbackOrder: fallbackOrder,
	}
}
